### PR TITLE
Heedls 871 learning content brand courses section

### PR DESCRIPTION
**Pull request by LukeJStuart**
### JIRA link 
[_HEEDLS-871_](https://softwiretech.atlassian.net/browse/HEEDLS-871?atlOrigin=eyJpIjoiM2ZiODFlYmM3OWZiNDI1ZGFkZmM1NjVhYzI2MzhlYWUiLCJwIjoiaiJ9)

### Description
Added ApplicationWithSections model to store application details as well as all of the relevant sections that will be needed for the courses display. Added a new section service to get all sections with tutorials for a given application and return the tutorials as well as the sections. Added new course data services to get applications by brand ID and to get the total numbers of recent progress records for each application for a given brand. Used these in a new course service to return a collection of ApplicationWithSections populated with all details necessary for population of the courses section of the learning content page. Added placeholder content to view to display course cards - sorting, filtering, and pagination has not yet been implemented.

### Screenshots
Tables do not format properly on mobile or tablet in Chrome developer mode - this needs to be addressed. Below is a screenshot for desktop.
localhost_5001_LearningContent_1 (4)

-----
### Developer checks
(Leave tasks unticked if they haven't been appropriate for your ticket.)

I have:
- [x] Run the formatter and made sure there are no IDE errors.
- [x] Written tests for the changes (accessibility tests, unit tests for controller, data services, services, view models, etc)
- [x] Manually tested my work with and without JavaScript. Full manual testing guidelines can be found here: https://softwiretech.atlassian.net/wiki/spaces/HEE/pages/6703648740/Testing
- [ ] Updated/added documentation in Swiki and/or Readme. Links (if any) are below:
- [ ] Updated my Jira ticket with information about other parts of the system that were touched as part of the MR and have to be sanity tested to ensure nothing’s broken.
- [x] Scanned over my own MR to ensure everything is as expected.

### CODE REVIEW

File: DigitalLearningSolutions.Data/Services/CourseService.cs [hunk 1]
```diff
@@ -405,6 +413,28 @@ Customisation customisation
...
+                application => new ApplicationWithSections(
+                    application,
+                    sectionService.GetSectionsThatHaveTutorialsAndPopulateTutorialsForApplication(
+                        application.ApplicationId
+                    ),
+                    maxPopularity == 0
+                        ? 0
+                        : (numRecordsByApplicationId.ContainsKey(application.ApplicationId)
+                            ? numRecordsByApplicationId[application.ApplicationId] / maxPopularity
+                            : 0)
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data/Services/CourseService.cs:432]: Could we make this a bit clearer by doing this in separate stages? Nested ternary statements are a bit hard to follow.

File: DigitalLearningSolutions.Data/Models/Courses/ApplicationWithSections.cs [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+            CategoryName = applicationDetails.CategoryName;
+            CourseTopicId = applicationDetails.CourseTopicId;
+            CourseTopic = applicationDetails.CourseTopic;
+            PLAssess = applicationDetails.PLAssess;
+            DiagAssess = applicationDetails.DiagAssess;
+            Sections = sections;
+            TotalMins = 0;
+            foreach (var section in Sections)
+            {
+                foreach (var tutorial in section.Tutorials)
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data/Models/Courses/ApplicationWithSections.cs:24]: You could do this in a neater way using IEnumerable.Sum()

File: DigitalLearningSolutions.Data/DataServices/CourseDataService.cs [hunk 3]
```diff
@@ -659,6 +684,28 @@ public void UpdateCourseOptions(CourseOptions courseOptions, int customisationId
...
+                @"SELECT Applications.ApplicationID, COUNT(Progress.ProgressID) AS Num_Recent_Progress_Records
+                    FROM Applications
+                        INNER JOIN Customisations ON Applications.ApplicationID = Customisations.ApplicationID
+                        INNER JOIN Progress ON Customisations.CustomisationID = Progress.CustomisationID
+                    WHERE Applications.BrandID = @brandId
+                      AND Applications.Debug = 0
+                      AND Applications.ArchivedDate IS NULL
+                      AND Progress.SubmittedTime > @threeMonthsAgo
+                    GROUP BY Applications.ApplicationID
+                    ORDER BY Num_Recent_Progress_Records DESC",
```
Comment (SteveJacksonSoft) [inline DigitalLearningSolutions.Data/DataServices/CourseDataService.cs:699]: Can we move the ordering into the service? I don't think we order queries anywhere else, so it would be helpful to keep a consistent approach of "This will be ordered by id or not at all" in data service queries, so our services know what they can and can't expect.

**Review by SteveJacksonSoft — CHANGES_REQUESTED** [review]

**Comment by SteveJacksonSoft** [comment]
Very neat code, nice to read!

File: DigitalLearningSolutions.Data/Services/CourseService.cs (hunk 1 above)
Comment (LukeJStuart) [reply DigitalLearningSolutions.Data/Services/CourseService.cs:432]: I've realised I am able to do away with the second statement by using .GetValueOrDefault()

**Comment by AlexJacksonDS** [comment]
Updated screenshots:
image
image
image

File: DigitalLearningSolutions.Web/Scripts/searchSortFilterAndPaginate/sort.ts [hunk 4]
```diff
@@ -1,6 +1,7 @@
 import moment from 'moment';
 import * as _ from 'lodash';
 import { ISearchableElement } from './searchSortFilterAndPaginate';
+import { search } from './search';
```
Comment (GeorgeBlackwell) [inline DigitalLearningSolutions.Web/Scripts/searchSortFilterAndPaginate/sort.ts:4]: Looks like this can be removed

File: DigitalLearningSolutions.Web/ViewModels/LearningContent/AllBrandCoursesViewModel.cs [hunk 5]
```diff
@@ -0,0 +1,23 @@
+﻿namespace DigitalLearningSolutions.Web.ViewModels.LearningContent
+{
+    using System.Collections.Generic;
+    using System.Linq;
+    using DigitalLearningSolutions.Data.Models.Courses;
+    using DigitalLearningSolutions.Web.Extensions;
+    using DigitalLearningSolutions.Web.ViewModels.Common.SearchablePage;
+    using DigitalLearningSolutions.Web.ViewModels.TrackingSystem.CourseSetup;
```
Comment (GeorgeBlackwell) [inline DigitalLearningSolutions.Web/ViewModels/LearningContent/AllBrandCoursesViewModel.cs:8]: This isn't being used

**Review by GeorgeBlackwell — CHANGES_REQUESTED** [review]
Nice work, just a couple of redundant using statements

File: DigitalLearningSolutions.Web/Views/LearningContent/_BrandCourseCard.cshtml [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+      </div>
+      <span hidden data-filter-value="@Model.CategoryFilter">
+        @Model.CategoryName
+      </span>
+      <span hidden data-filter-value="@Model.TopicFilter">
+        @Model.CourseTopic
+      </span>
+      <span hidden name="created-date">
+        @Model.CreatedDate
+      </span>
```
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Web/Views/LearningContent/_BrandCourseCard.cshtml:32]: Feels a bit weird to be sorting by something that is hidden. Do we want to display this instead?

Also could use "data-name-for-sorting" as the attribute for the sortable name.

File: DigitalLearningSolutions.Web/Scripts/searchSortFilterAndPaginate/searchSortFilterAndPaginate.ts [hunk 7]
```diff
@@ -256,8 +260,12 @@ export class SearchSortFilterAndPaginate {
     return newResultCountMessage;
   }
 
-  private static scrollToTop(): void {
-    window.scrollTo(0, 0);
+  private scrollToTop(): void {
+    if (this.idConsideredTopForScrolling !== '') {
+      document.getElementById(this.idConsideredTopForScrolling)?.scrollIntoView();
```
Comment (DanBloxham-sw) [inline DigitalLearningSolutions.Web/Scripts/searchSortFilterAndPaginate/searchSortFilterAndPaginate.ts:265]: What's this new change for?

**Review by DanBloxham-sw — CHANGES_REQUESTED** [review]
Think I've covered everything, number of small points and questions.

File: DigitalLearningSolutions.Web/DigitalLearningSolutions.Web.csproj [hunk 8]
```diff
@@ -1,148 +1,154 @@
 <Project Sdk="Microsoft.NET.Sdk.Web">
 
-  <PropertyGroup>
-    <TargetFramework>netcoreapp3.1</TargetFramework>
-    <Nullable>annotations</Nullable>
-  </PropertyGroup>
+    <PropertyGroup>
```
[the remark this answers is not in the document]
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Web/DigitalLearningSolutions.Web.csproj:3]: No, this is just something that happened automatically, not sure why

File: DigitalLearningSolutions.Data/DataServices/CourseDataService.cs [hunk 9]
```diff
@@ -659,6 +685,27 @@ public void UpdateCourseOptions(CourseOptions courseOptions, int customisationId
             );
         }
 
+        public Dictionary<int, int> GetNumsOfRecentProgressRecordsForBrand(int brandId, DateTime threeMonthsAgo)
```
[the remark this answers is not in the document]
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Data/DataServices/CourseDataService.cs:688]: The old site is doing
```
(
           SELECT COUNT(*) AS Expr1
           FROM Progress AS p
               INNER JOIN Customisations AS c
                   ON p.CustomisationID = c.CustomisationID
           WHERE (c.ApplicationID = a.ApplicationID)
                 AND (p.SubmittedTime > DATEADD(M, -3, GETDATE()))
       ) AS Popularity
```
I'm not really sure why we aren't doing this in a single query in the new version, but it looks approximately the same to me with no filtering of the progress records?

File: DigitalLearningSolutions.Data/Models/Courses/ApplicationDetails.cs [hunk 10]
```diff
@@ -1,9 +1,24 @@
 ﻿namespace DigitalLearningSolutions.Data.Models.Courses
 {
+    using System;
     using DigitalLearningSolutions.Data.Models.SearchSortFilterPaginate;
 
     public class ApplicationDetails : BaseSearchableItem
     {
+        public ApplicationDetails(){}
```
[the remark this answers is not in the document]
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Data/Models/Courses/ApplicationDetails.cs:8]: Yes, for various SQL queries.

File: DigitalLearningSolutions.Data/Services/CourseService.cs [hunk 11]
```diff
@@ -405,6 +413,32 @@ Customisation customisation
...
         }
 
+        public IEnumerable<ApplicationWithSections> GetApplicationsThatHaveSectionsByBrandId(int brandId)
+        {
+            var numRecordsByApplicationId =
+                courseDataService.GetNumsOfRecentProgressRecordsForBrand(brandId, clockService.UtcNow.AddMonths(-3));
+
+            var applications = courseDataService.GetApplicationsByBrandId(brandId);
+
+            double maxPopularity = numRecordsByApplicationId.Any() ? numRecordsByApplicationId.Values.Max() : 0;
```
[the remark this answers is not in the document]
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Data/Services/CourseService.cs:423]: I think so yeah

File: DigitalLearningSolutions.Web/Views/LearningContent/Index.cshtml [hunk 12]
```diff
@@ -29,9 +29,33 @@
...
-  <p class="nhsuk-u-margin-top-4">Course Tutorials</p>
+  <h2 class="nhsuk-heading-l">Course Tutorials</h2>
   <p class="nhsuk-body-m">Access to this content does not require login and progress is not tracked.</p>
   foreach (var tutorial in Model.Tutorials) {
     <partial name="_TutorialCard" model="tutorial" />
   }
 }
+
+@if (Model.Applications.Any()) {
+  <div id="@(Model.JavascriptSearchSortFilterPaginateEnabled ? "js-styling-hidden-area-while-loading" : "no-js-styling")">
```
[the remark this answers is not in the document]
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Web/Views/LearningContent/Index.cshtml:40]: Yes, since the page has content above the paginated area that is outside the paginated area, and the paginated area may not always be included, I didn't include it as they overlap, and I don't think there will be a good way to guarantee the spinner always displays below the dynamic tutorials list.

File: DigitalLearningSolutions.Web/Scripts/searchSortFilterAndPaginate/searchSortFilterAndPaginate.ts (hunk 7 above)
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Web/Scripts/searchSortFilterAndPaginate/searchSortFilterAndPaginate.ts:265]: The searchable/paginated area is below a big chunk of content, and when we interact with various controls, it scrolls to top. This allows us to define a different element at the 'top' that it scrolls to. In this case the Courses heading

File: DigitalLearningSolutions.Web/Views/LearningContent/Index.cshtml [hunk 13]
```diff
@@ -29,9 +29,33 @@
...
+    <h2 class="nhsuk-heading-l nhsuk-u-padding-top-7" id="courses-heading">Courses</h2>
+    <partial name="SearchablePage/Configurations/_ThreeQuarterWidthSortAndFilter" model="Model" />
```
[the remark this answers is not in the document]
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Web/Views/LearningContent/Index.cshtml:42]: This was the only place left using this, so I've just renamed to remove the width

File: DigitalLearningSolutions.Data/DataServices/CourseDataService.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Data/DataServices/CourseDataService.cs:688]: Confirmed with Kevin, no filtering

File: DigitalLearningSolutions.Web/Views/LearningContent/_BrandCourseCard.cshtml (hunk 6 above)
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Web/Views/LearningContent/_BrandCourseCard.cshtml:32]: Confirmed with Kevin, we want to add this as visible

**Review by DanBloxham-sw — CHANGES_REQUESTED** [review]
Mostly a few question to check that everything works as we expect.

File: DigitalLearningSolutions.Web/Views/LearningContent/_BrandCourseCard.cshtml [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+      </div>
+      <span hidden data-filter-value="@Model.CategoryFilter">
+        @Model.CategoryName
+      </span>
+      <span hidden data-filter-value="@Model.TopicFilter">
+        @Model.CourseTopic
+      </span>
+      <span class="nhsuk-u-margin-bottom-4" data-name-for-sorting="created-date">
+        <span class="nhsuk-u-font-weight-bold">Created</span> @Model.CreatedDate.ToString(DateHelper.StandardDateFormat)
+      </span>
```
[the remark this answers is not in the document]
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Web/Views/LearningContent/_BrandCourseCard.cshtml:33]: moment that we're using for the parsing seems smart enough to deal with this, but we have lost the time sorting here. Will tidy this up.

File: DigitalLearningSolutions.Web/Views/LearningContent/Index.cshtml [hunk 15]
```diff
@@ -29,9 +29,33 @@
...
+
+    <partial name="SearchablePage/_Pagination" model="Model" />
+  </div>
+}
+
+<input type="hidden" id="brand-id" asp-for="BrandId" />
+@if (Model.JavascriptSearchSortFilterPaginateEnabled) {
+@section scripts {
+  <script src="@Url.Content("~/js/learningContent/brands.js")" asp-append-version="true"></script>
+}}
```
[the remark this answers is not in the document]
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Web/Views/LearningContent/Index.cshtml:61]: We don't but this should probably be pulled inside the if applications.Any check. No need to load the javascript unnecessarily

File: DigitalLearningSolutions.Web/Views/LearningContent/Index.cshtml (hunk 12 above)
[the remark this answers is not in the document]
Comment (AlexJacksonDS) [reply DigitalLearningSolutions.Web/Views/LearningContent/Index.cshtml:40]: If I move it outside, then won't I not be able to have the conditional Courses section due to the page structure?
